File: sdk/python/src/telemetry_agent/_params.py
"""Request attribute extraction: turn a raw request into the ``params`` array of a record.

An error or latency record is only actionable if it says which inputs the handler saw,
so every location is described: query string, body (JSON, form, multipart), path
variables, cookies and the headers a handler may key behaviour off. Everything is
parsed from the raw bytes rather than from the framework's parsed view, because the
framework only exposes what the handler asked for, and the interesting requests are
usually the ones where the handler asked for the wrong thing.

Values are described as ``(sha256, length, 256-character sample)`` rather than stored
verbatim: the hash makes it possible to group requests that carried the same value, and
to tell a default value from an unusual one, without keeping user data in the pipeline.
"""

from __future__ import annotations

import hashlib
import json
import re
from typing import Any, Iterator
from urllib.parse import parse_qsl

SAMPLE_MAX = 256

# Nesting and volume guards. Documents of a few megabytes with deep nesting do arrive;
# extraction runs on the request path, so its cost is bounded by construction.
JSON_DEPTH_MAX = 16
DEFAULT_MAX_PARAMS = 1024

# Headers worth describing: the ones a handler, a proxy or a cache may key behaviour
# off. Everything ``x-*`` is included because custom headers are where per-tenant and
# feature-toggle routing lives.
DESCRIBED_HEADERS = frozenset(
    {
        "host",
        "referer",
        "user-agent",
        "origin",
        "content-type",
        "accept-language",
        "authorization",
        "forwarded",
        "true-client-ip",
    }
)


def is_described_header(name: str) -> bool:
    lowered = name.lower()
    return lowered in DESCRIBED_HEADERS or lowered.startswith("x-")


def sha256_of(value: str | bytes) -> str:
    if isinstance(value, str):
        # surrogatepass keeps lone surrogates (which percent-decoding can produce)
        # hashable instead of raising on the request path.
        value = value.encode("utf-8", "surrogatepass")
    return hashlib.sha256(value).hexdigest()


def _to_text(value: str | bytes) -> str:
    if isinstance(value, str):
        return value
    return value.decode("utf-8", "replace")


def describe_param(name: str, location: str, value: str | bytes) -> dict[str, Any]:
    """Describe one input. ``value_len`` counts bytes for binary values."""
    text = _to_text(value)
    return {
        "name": name,
        "in": location,
        "value_sha256": sha256_of(value),
        "value_len": len(value) if isinstance(value, bytes) else len(text),
        "sample": text[:SAMPLE_MAX],
    }


def json_scalar_to_text(value: Any) -> str:
    """Render a JSON leaf the way it looked on the wire.

    ``"laptop"`` must hash to sha256(b"laptop"), and a number must hash like its
    textual form, so that the same value carried as JSON, as a form field or as a query
    parameter groups together. Hence: strings raw, everything else via json.dumps
    (``true`` / ``null`` / ``1001``).
    """
    if isinstance(value, str):
        return value
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    try:
        return json.dumps(value)
    except (TypeError, ValueError):
        return str(value)


def flatten_json(value: Any, prefix: str = "", depth: int = 0) -> Iterator[tuple[str, str]]:
    """Yield ``(dotted_path, text_value)`` for every leaf of a decoded JSON document.

    ``{"filter": {"tags": ["a"]}}`` yields ``filter.tags.0``. Empty containers are
    yielded as leaves so their *name* still appears as an observed input.
    """
    if depth > JSON_DEPTH_MAX:
        return
    if isinstance(value, dict):
        if not value:
            yield (prefix or "body", "{}")
            return
        for key, sub in value.items():
            path = f"{prefix}.{key}" if prefix else str(key)
            yield from flatten_json(sub, path, depth + 1)
    elif isinstance(value, (list, tuple)):
        if not value:
            yield (prefix or "body", "[]")
            return
        for index, sub in enumerate(value):
            path = f"{prefix}.{index}" if prefix else str(index)
            yield from flatten_json(sub, path, depth + 1)
    else:
        yield (prefix or "body", json_scalar_to_text(value))


_CONTENT_DISPOSITION_NAME = re.compile(rb'name="((?:[^"\\]|\\.)*)"', re.IGNORECASE)
_CONTENT_DISPOSITION_FILENAME = re.compile(rb'filename="((?:[^"\\]|\\.)*)"', re.IGNORECASE)


def _multipart_boundary(content_type: str) -> bytes | None:
    for chunk in content_type.split(";")[1:]:
        key, _, raw = chunk.strip().partition("=")
        if key.strip().lower() == "boundary":
            value = raw.strip().strip('"')
            if value:
                return value.encode("latin-1", "replace")
    return None


def iter_multipart(body: bytes, content_type: str) -> Iterator[tuple[str, str | bytes]]:
    """Yield ``(field_name, value)`` for a multipart body, parsed from raw bytes.

    Written by hand rather than with a full parser because this one must survive the
    malformed bodies that show up in production (missing terminator, truncated upload,
    bogus part headers) without raising: an exception here would cost the whole record,
    and malformed uploads are exactly the requests someone will want to look at.
    File parts additionally yield ``<field>.filename``.
    """
    boundary = _multipart_boundary(content_type)
    if not boundary:
        return
    for part in body.split(b"--" + boundary):
        if not part or part in (b"--", b"--\r\n"):
            continue
        head, sep, payload = part.partition(b"\r\n\r\n")
        if not sep:
            continue
        name_match = _CONTENT_DISPOSITION_NAME.search(head)
        if not name_match:
            continue
        name = name_match.group(1).decode("utf-8", "replace")
        payload = payload[:-2] if payload.endswith(b"\r\n") else payload
        file_match = _CONTENT_DISPOSITION_FILENAME.search(head)
        if file_match:
            yield (f"{name}.filename", file_match.group(1).decode("utf-8", "replace"))
        yield (name, payload)


def parse_cookie_header(raw: str) -> Iterator[tuple[str, str]]:
    """Split a Cookie header by hand.

    ``http.cookies.SimpleCookie`` silently discards pairs it considers illegal, and a
    malformed cookie is usually the reason the request is being looked at in the first
    place, so it cannot be used here.
    """
    for chunk in raw.split(";"):
        chunk = chunk.strip()
        if not chunk:
            continue
        name, sep, value = chunk.partition("=")
        name = name.strip()
        if name:
            yield (name, value.strip() if sep else "")


def graphql_params(query: Any, variables: Any, operation_name: Any) -> Iterator[tuple[str, str]]:
    """Yield ``in="graphql"`` attributes for one GraphQL operation.

    Variables keep a ``variables.`` prefix so a dashboard can name one argument
    (``variables.id``) with no ambiguity against ``query`` itself. The document is
    reported verbatim rather than parsed: turning a document into a selection set is
    interpretation, and interpretation belongs downstream, not on the request path.
    """
    if isinstance(query, str) and query:
        yield ("query", query)
    if isinstance(operation_name, str) and operation_name:
        yield ("operationName", operation_name)
    if isinstance(variables, (dict, list)):
        yield from ((f"variables.{path}", text) for path, text in flatten_json(variables, "") if path)
    elif isinstance(variables, str) and variables:
        yield ("variables", variables)


class ParamCollector:
    """Ordered, de-duplicated, bounded accumulator of described inputs.

    De-duplication is on (location, name, value hash) rather than (location, name): a
    repeated parameter carrying a *different* value is the interesting case -- it is
    what makes two identical-looking requests behave differently -- and collapsing it
    to a single entry would hide it. Identical repeats do collapse.
    """

    __slots__ = ("_entries", "_seen", "_max", "truncated")

    def __init__(self, max_params: int = DEFAULT_MAX_PARAMS) -> None:
        self._entries: list[dict[str, Any]] = []
        self._seen: set[tuple[str, str, str]] = set()
        self._max = max_params
        self.truncated = False

    def add(self, name: str, location: str, value: str | bytes) -> None:
        if len(self._entries) >= self._max:
            self.truncated = True
            return
        entry = describe_param(name, location, value)
        key = (location, entry["name"], entry["value_sha256"])
        if key in self._seen:
            return
        self._seen.add(key)
        self._entries.append(entry)

    def add_many(self, pairs, location: str) -> None:
        for name, value in pairs:
            self.add(name, location, value)

    def extend_entries(self, entries) -> None:
        """Merge already-described inputs (from the graphql/websocket helpers)."""
        for entry in entries:
            key = (entry.get("in", ""), entry.get("name", ""), entry.get("value_sha256", ""))
            if key in self._seen or len(self._entries) >= self._max:
                self.truncated = self.truncated or len(self._entries) >= self._max
                continue
            self._seen.add(key)
            self._entries.append(entry)

    @property
    def entries(self) -> list[dict[str, Any]]:
        return self._entries


def collect_query(collector: ParamCollector, query_string: str | bytes) -> None:
    text = _to_text(query_string)
    if not text:
        return
    try:
        pairs = parse_qsl(text, keep_blank_values=True)
    except ValueError:
        collector.add("query_string", "raw", text)
        return
    collector.add_many(pairs, "query")


def collect_headers(collector: ParamCollector, headers) -> None:
    """``headers`` is an iterable of (name, value) with str values."""
    for name, value in headers:
        lowered = name.lower()
        if lowered == "cookie":
            collector.add_many(parse_cookie_header(value), "cookie")
            continue
        if is_described_header(lowered):
            collector.add(lowered, "header", value)


def collect_body(collector: ParamCollector, body: bytes, content_type: str) -> None:
    """Describe the body by content type, sniffing JSON when the type is missing."""
    if not body:
        return
    base = (content_type or "").split(";")[0].strip().lower()
    if base == "application/json" or base.endswith("+json") or (not base and body[:1] in (b"{", b"[")):
        try:
            decoded = json.loads(body)
        except (ValueError, UnicodeDecodeError):
            collector.add("body", "raw", body)
            return
        collector.add_many(flatten_json(decoded), "json")
        if isinstance(decoded, dict) and isinstance(decoded.get("query"), str):
            # A GraphQL call is JSON on the wire; describing it under both locations
            # lets a dashboard be written against either view of the same request.
            collector.add_many(
                graphql_params(decoded.get("query"), decoded.get("variables"), decoded.get("operationName")),
                "graphql",
            )
        return
    if base == "application/x-www-form-urlencoded":
        try:
            pairs = parse_qsl(body.decode("utf-8", "replace"), keep_blank_values=True)
        except ValueError:
            collector.add("body", "raw", body)
            return
        collector.add_many(pairs, "body")
        return
    if base.startswith("multipart/"):
        found = False
        for name, value in iter_multipart(body, content_type):
            found = True
            collector.add(name, "multipart", value)
        if not found:
            collector.add("body", "raw", body)
        return
    if base in ("application/graphql", "application/graphql+json"):
        collector.add("query", "graphql", body)
        return
    collector.add("body", "raw", body)

File: sdk/php/src/Params.php
<?php

declare(strict_types=1);

namespace Internal\Telemetry;

/**
 * Turn a raw request into the `params` array of a record.
 *
 * A latency or error record is only actionable if it says which inputs the handler
 * could see, so every location is described: query string, form body, JSON body,
 * multipart field names, path variables, cookies, and the headers a handler may key
 * behaviour off.
 *
 * Everything is parsed from the RAW bytes rather than from PHP's parsed view. That is
 * not fussiness. `$_GET`, `$_POST` and `$_COOKIE` are lossy in three ways that matter
 * here: a repeated name keeps only its last value, a dot or a space in a name is
 * rewritten to an underscore, and a name with `[]` in it becomes an array. The
 * requests worth looking at are exactly the ones where those details differ from what
 * the handler thought it received.
 *
 * Values are described as (sha256, length, 256-character sample) rather than kept
 * verbatim: the hash groups requests that carried the same value, and separates a
 * default value from an unusual one, without carrying user data down the pipeline.
 */
final class Params
{
    /**
     * Headers worth describing: the ones a handler, a proxy or a cache may key
     * behaviour off. Everything `x-*` is included because custom headers are where
     * per-tenant and feature-toggle routing lives.
     */
    private const DESCRIBED_HEADERS = [
        'host' => true,
        'referer' => true,
        'user-agent' => true,
        'origin' => true,
        'content-type' => true,
        'accept-language' => true,
        'authorization' => true,
        'forwarded' => true,
        'true-client-ip' => true,
    ];

    /** Guard on nesting. Documents a few megabytes deep do arrive, and this runs on the request path. */
    private const JSON_DEPTH_MAX = 16;

    public static function isDescribedHeader(string $name): bool
    {
        $lowered = strtolower($name);

        return isset(self::DESCRIBED_HEADERS[$lowered]) || str_starts_with($lowered, 'x-');
    }

    public static function sha256(string $value): string
    {
        return hash('sha256', $value);
    }

    /**
     * Describe one input.
     *
     * `value_len` counts bytes. PHP strings are byte strings and an upload part is
     * binary, so bytes is the only measure that is meaningful for all of them.
     *
     * @return array{name:string,in:string,value_sha256:string,value_len:int,sample:string}
     */
    public static function describe(string $name, string $location, string $value, int $sampleMax = Config::SAMPLE_MAX): array
    {
        return [
            'name' => $name,
            'in' => $location,
            'value_sha256' => self::sha256($value),
            'value_len' => strlen($value),
            'sample' => self::clip($value, $sampleMax),
        ];
    }

    /**
     * Cut a value to a sample without splitting a UTF-8 sequence down the middle.
     *
     * A split sequence is not merely ugly: it makes the record unencodable, and losing
     * a whole record to one ragged byte is the expensive failure here.
     */
    public static function clip(string $value, int $max = Config::SAMPLE_MAX): string
    {
        if ($max <= 0 || strlen($value) <= $max) {
            return $value;
        }
        $cut = substr($value, 0, $max);
        // Walk back over at most three continuation bytes, then drop the lead byte if
        // the sequence it opened did not fit.
        $back = 0;
        while ($back < 3 && $back < strlen($cut)) {
            $byte = ord($cut[strlen($cut) - 1 - $back]);
            if ($byte >= 0x80 && $byte <= 0xBF) {
                $back++;
                continue;
            }
            if ($byte >= 0xC0) {
                $needed = match (true) {
                    $byte >= 0xF0 => 4,
                    $byte >= 0xE0 => 3,
                    default => 2,
                };
                if ($needed > $back + 1) {
                    return substr($cut, 0, strlen($cut) - 1 - $back);
                }
            }
            break;
        }

        return $cut;
    }

    /**
     * Split a urlencoded string into ordered (name, value) pairs, duplicates kept.
     *
     * `parse_str()` cannot be used: it collapses `a=1&a=2` to a single entry and
     * rewrites names, which is precisely the information this record exists to carry.
     *
     * @return list<array{0:string,1:string}>
     */
    public static function parseQuery(string $raw): array
    {
        $out = [];
        if ($raw === '') {
            return $out;
        }
        foreach (explode('&', $raw) as $chunk) {
            if ($chunk === '') {
                continue;
            }
            $position = strpos($chunk, '=');
            if ($position === false) {
                $name = urldecode($chunk);
                $value = '';
            } else {
                $name = urldecode(substr($chunk, 0, $position));
                $value = urldecode(substr($chunk, $position + 1));
            }
            if ($name === '') {
                continue;
            }
            $out[] = [$name, $value];
        }

        return $out;
    }

    /**
     * Split a Cookie header by hand.
     *
     * `$_COOKIE` drops pairs PHP considers illegal and rewrites the names of the rest.
     * A malformed cookie is usually the reason the request is being looked at.
     *
     * @return list<array{0:string,1:string}>
     */
    public static function parseCookieHeader(string $raw): array
    {
        $out = [];
        foreach (explode(';', $raw) as $chunk) {
            $chunk = trim($chunk);
            if ($chunk === '') {
                continue;
            }
            $position = strpos($chunk, '=');
            if ($position === false) {
                $out[] = [$chunk, ''];
                continue;
            }
            $name = trim(substr($chunk, 0, $position));
            if ($name === '') {
                continue;
            }
            $out[] = [$name, trim(substr($chunk, $position + 1))];
        }

        return $out;
    }

    /**
     * Render a JSON leaf the way it looked on the wire.
     *
     * `"laptop"` must hash to sha256('laptop'), and a number must hash like its textual
     * form, so that the same value carried as JSON, as a form field or as a query
     * parameter groups together.
     */
    public static function jsonScalarToText(mixed $value): string
    {
        if (is_string($value)) {
            return $value;
        }
        if ($value === null) {
            return 'null';
        }
        if (is_bool($value)) {
            return $value ? 'true' : 'false';
        }
        $encoded = json_encode($value, JSON_UNESCAPED_SLASHES | JSON_INVALID_UTF8_SUBSTITUTE);

        return $encoded === false ? (string) $value : $encoded;
    }

    /**
     * Flatten a decoded JSON document into ordered (dotted path, text) pairs.
     *
     * `{"filter":{"tags":["a"]}}` yields `filter.tags.0`. Empty containers are yielded
     * as leaves so their NAME still appears as an observed input.
     *
     * @return list<array{0:string,1:string}>
     */
    public static function flattenJson(mixed $value, string $prefix = '', int $depth = 0): array
    {
        if ($depth > self::JSON_DEPTH_MAX) {
            return [];
        }
        if (is_array($value)) {
            if ($value === []) {
                return [[$prefix === '' ? 'body' : $prefix, array_is_list($value) ? '[]' : '{}']];
            }
            $out = [];
            foreach ($value as $key => $sub) {
                $path = $prefix === '' ? (string) $key : $prefix . '.' . $key;
                foreach (self::flattenJson($sub, $path, $depth + 1) as $pair) {
                    $out[] = $pair;
                }
            }

            return $out;
        }
        if (is_object($value)) {
            return self::flattenJson(get_object_vars($value), $prefix, $depth);
        }

        return [[$prefix === '' ? 'body' : $prefix, self::jsonScalarToText($value)]];
    }

    /**
     * Header name/value pairs derived from the CGI-style server array.
     *
     * `getallheaders()` is not used: it only exists under some SAPIs, and under those
     * it hands back a normalised view. Deriving from HTTP_* works everywhere PHP runs.
     *
     * @param  array<string,mixed>              $server
     * @return list<array{0:string,1:string}>
     */
    public static function headersFromServer(array $server): array
    {
        $out = [];
        foreach ($server as $key => $value) {
            if (!is_string($key) || !is_string($value)) {
                continue;
            }
            if (str_starts_with($key, 'HTTP_')) {
                $out[] = [strtolower(str_replace('_', '-', substr($key, 5))), $value];
                continue;
            }
            if ($key === 'CONTENT_TYPE') {
                $out[] = ['content-type', $value];
            } elseif ($key === 'CONTENT_LENGTH') {
                $out[] = ['content-length', $value];
            }
        }

        return $out;
    }

    /**
     * Base media type of a Content-Type header, lower-cased and without parameters.
     */
    public static function baseContentType(string $contentType): string
    {
        $base = explode(';', $contentType, 2)[0];

        return strtolower(trim($base));
    }
}
